File: selected_coding_interview/codes/java/lc_946_validate_stack_sequences/lc_946_validate_stack_sequences.java
/*
* File: lc_946_validate_stack_sequences.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_946_validate_stack_sequences;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public boolean validateStackSequences(int[] pushed, int[] popped) {
            Stack<Integer> stack = new Stack<>();
            int i = 0;
            for (int num : pushed) {
                stack.push(num); // num 入栈
                while (!stack.isEmpty() && stack.peek() == popped[i]) { // 循环判断与出栈
                    stack.pop();
                    i++;
                }
            }
            return stack.isEmpty();
        }
    }

public class lc_946_validate_stack_sequences {
    public static void main(String[] args) {
        // ======= Test Case =======
        int[] test_input_pushed = new int[]{1, 2, 3, 4, 5};
        int[] test_input_popped = new int[]{4, 5, 3, 2, 1};
        var expected_output = true;

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.validateStackSequences(test_input_pushed, test_input_popped);
        System.out.println(result);

    }
}

File: selected_coding_interview/codes/python/lc_946_validate_stack_sequences.py
"""
File: lc_946_validate_stack_sequences.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack, i = [], 0
        for num in pushed:
            stack.append(num) # num 入栈
            while stack and stack[-1] == popped[i]: # 循环判断与出栈
                stack.pop()
                i += 1
        return not stack


# ======= Test Case =======
test_input_pushed = [1, 2, 3, 4, 5]
test_input_popped = [4, 5, 3, 2, 1]
expected_output = True

# ====== Driver Code ======
slt = Solution()
result = slt.validateStackSequences(test_input_pushed, test_input_popped)
print(result)
